backend/de.metas.fresh/de.metas.fresh.base/src/main/sql/postgresql/ddl/functions/Fresh_Account_Info_Report.sql

```diff
@@ -337,121 +100,481 @@ CREATE OR REPLACE FUNCTION report.fresh_account_info_report(
+                 END
+
+             AND fa.ad_org_id = $11
+
+             AND CASE
+                     WHEN ($10 = 'N') THEN
+                             1 = 1
+                                      ELSE
+                         (
+                             fa.C_Activity_ID IS NULL
+                             )
+                 END
+
+           ORDER BY fa.Param_Acct_Value,
+                    fa.Account_ID,
+                    fa.DateAcct::Date,
+                    fa.Fact_Acct_ID;
+    END INIT_TABLE;
+
+    <<UPDATE_TABLE>>
+        DECLARE
+        currencies numeric[];
+        length int;
+        row report.tmp_fresh_Account_Info_Report_Sub%rowtype;
+        v_currencyid1 numeric;
+        v_currencyid2 numeric;
+        v_currencyid3 numeric;
+        v_currencyid4 numeric;
+        v_currencyid5 numeric;
+        v_currency1 text;
+        v_currency2 text;
+        v_currency3 text;
+        v_currency4 text;
+        v_currency5 text;
+        v_sourcebalance1 numeric;
+        v_sourcebalance2 numeric;
+        v_sourcebalance3 numeric;
+        v_sourcebalance4 numeric;
+        v_sourcebalance5 numeric;
+        v_rollingbalance1 numeric;
+        v_rollingbalance2 numeric;
+        v_rollingbalance3 numeric;
+        v_rollingbalance4 numeric;
+        v_rollingbalance5 numeric;
+    BEGIN
+        currencies := ARRAY (SELECT DISTINCT source_currency_id FROM report.tmp_fresh_Account_Info_Report_Sub);
+        length := array_length(currencies, 1);
+
+        IF length > 0 AND length <= 5 AND $9 = 'Y'THEN --p_showcurrencyexchange (used for option show foreign currency)
+
+            v_currencyid1 :=  currencies[1];
+            SELECT iso_code FROM c_currency WHERE c_currency_id = v_currencyid1 INTO v_currency1;
+            IF length > 1 THEN
+                v_currencyid2 :=  currencies[2];
+                SELECT iso_code FROM c_currency WHERE c_currency_id = v_currencyid2 INTO v_currency2;
+            END IF;
+            IF length > 2 THEN
+                v_currencyid3 :=  currencies[3];
+                SELECT iso_code FROM c_currency WHERE c_currency_id = v_currencyid3 INTO v_currency3;
+            END IF;
+            IF length > 3 THEN
+                v_currencyid4 :=  currencies[4];
+                SELECT iso_code FROM c_currency WHERE c_currency_id = v_currencyid4 INTO v_currency4;
+            END IF;
+            IF length > 4 THEN
+                v_currencyid5 :=  currencies[5];
+                SELECT iso_code FROM c_currency WHERE c_currency_id = v_currencyid5 INTO v_currency5;
+            END IF;
+
+            FOR row IN SELECT * FROM report.tmp_fresh_Account_Info_Report_Sub ORDER BY DateAcct, Fact_Acct_ID
+                LOOP
+                    IF row.source_currency_id = v_currencyid1 THEN
+                        v_sourcebalance1 := (de_metas_acct.sourceAcctBalanceToDate(row.Account_ID, row.C_AcctSchema_ID, row.start_date_acct, row.ad_org_id, v_currencyid1)).Balance;
```

Comment:
IMHO u shall call de_metas_acct.sourceAcctBalanceToDate only one time per currency because this is an expensive function.